[Async scrolling] Programmatic scrolls can cause missing tiles and offset hit-testing

smfr opened the pull request:
#### e8697834ad4b3d036032823f29b7d4329ffb45ec
<pre>
[Async scrolling] Programmatic scrolls can cause missing tiles and offset hit-testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=262287">https://bugs.webkit.org/show_bug.cgi?id=262287</a>
<a href="https://rdar.apple.com/116205365">rdar://116205365</a>

Reviewed by Alan Baradlay and Abrar Rahman Protyasha.

This change fixes some long-standing issues with asynchronous scrolling, where
UI and web processes could be left with different scroll positions when both
user and programmatic scrolling was happening at the same time. This could result
in bad tile coverage (missing content), and offset hit-testing.

The fundamental problem was that IPC message for programmatic scrolling, from
web to UI process, could overlap with messages in the other direction for user
scrolling; the messages would pass like ships in the night, with no final
reconciliation step.

307867@main introduced the concept of `ScrollRequestIdentifier`. This change
makes use of those identifiers to ensure that the web process is updated with
the correct scroll position via a message from the UI process back to the web
process. Every programmatic scroll IPC now triggers a response message, which
returns the originating identifier. This allows `RemoteScrollingCoordinator` in
the web process to know when there&apos;s programmatic scroll in flight. When there
is, we have to ignore user scrolls, because they will contain a scroll position
that is stale relative to the current web process scroll position. With multiple
programmatic scrolls in a row, we also have to ignore the response for the same
reason. We need to take care to fire `scrollend` when necessary.

ScrollUpdate now has a data member with two variants; one for responses to
programmatic scrolls (`ScrollRequestResponseData`) and one for other messages
from UI to web process.

We actually apply the scroll position in `RemoteScrollingCoordinator::scrollUpdateForNode()`,
to handle the case where the web process sent a &quot;delta&quot; scroll (e.g. for scroll
anchoring or `scrollBy`); the response will contain the most recent UI-side scroll
position with the delta applied.

There was already a `ProgrammaticScrollDidEnd` sent back to the web process; now
that we always send `ScrollRequestResponse` updates back, use that to trigger
`scrollend` via a `ShouldFireScrollEnd` member on ScrollUpdate.
`scrollingTreeNodeDidStopProgrammaticScroll` is no longer needed.

`scrollingTreeNodeRequestsScroll()` has to distinguish between &quot;scrolling tree
handled it for animated scroll&quot;, and &quot;scrolling tree delayed handling for
delegated scrolling&quot; so add `RequestsScrollHandling` to indicate that.

The changes in `ScrollingTreeScrollingNode::handleScrollPositionRequest()` and
`RemoteScrollingCoordinatorProxy::adjustMainFrameDelegatedScrollPosition()` are
needed to handle the iOS-specific behavior, seen in `RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction()`,
where the application of the `requestedScroll` is delayed until after layer tree
commit; `didHandleScrollRequestForNode()` needs to be called once the final scroll
position has been computed (this is tested by existing tests).

Tests: fast/scrolling/mac/momentum-then-programmatic-hit-test.html
       fast/scrolling/mac/momentum-then-programmatic-missing-tiles.html

* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-hit-test-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-hit-test.html: Added.
* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-missing-tiles-expected.html: Added.
* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-missing-tiles.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/dirty-contents-reselect-anchor.tentative-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::requestScrollToPosition):
(WebCore::AsyncScrollingCoordinator::synchronizeStateFromScrollingTree):
(WebCore::AsyncScrollingCoordinator::applyScrollPositionUpdate):
(WebCore::AsyncScrollingCoordinator::updateScrollPositionAfterAsyncScroll):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::ScrollUpdate::canMerge const):
(WebCore::ScrollUpdate::merge):
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::ScrollUpdate::canMerge const): Deleted.
(WebCore::ScrollUpdate::merge): Deleted.
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeDidStopWheelEventScroll):
(WebCore::ScrollingTree::scrollingTreeNodeRequestsScroll):
(WebCore::ScrollingTree::didHandleScrollRequestForNode):
(WebCore::ScrollingTree::scrollingTreeNodeDidStopProgrammaticScroll): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::handleScrollPositionRequest):
(WebCore::ScrollingTreeScrollingNode::didStopProgrammaticScroll): Deleted.
* Source/WebCore/page/scrolling/ThreadedScrollingTree.cpp:
(WebCore::ThreadedScrollingTree::scrollingTreeNodeRequestsScroll):
(WebCore::ThreadedScrollingTree::scrollingTreeNodeDidScroll):
(WebCore::ThreadedScrollingTree::didHandleScrollRequestForNode):
(WebCore::ThreadedScrollingTree::scrollingTreeNodeScrollUpdated):
* Source/WebCore/page/scrolling/ThreadedScrollingTree.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::adjustMainFrameDelegatedScrollPosition):
(WebKit::RemoteScrollingCoordinatorProxy::sendScrollingTreeNodeUpdate):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeRequestsScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopAnimatedScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopWheelEventScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeRequestsScroll):
(WebKit::RemoteScrollingTree::didHandleScrollRequestForNode):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopProgrammaticScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::scrollingTreeNodeRequestsScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopAnimatedScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopWheelEventScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeRequestsScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopProgrammaticScroll): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::willSendScrollPositionRequest):
(WebKit::RemoteScrollingCoordinator::scrollUpdateForNode):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):

Canonical link: <a href="https://commits.webkit.org/308215@main">https://commits.webkit.org/308215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8eaafaf67d16c70cc4b4bacabaa35d6b59592e4b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/146792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/19473 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/168/builds/13005 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/155460 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/100181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/a8b1e4e7-c423-484b-89da-162335e9a431) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/148667 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/19932 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/19374 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/113104 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/100181 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/8e6baf8c-058b-4cfe-8bd7-48f55b825556) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/149755 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/15361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/131903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/93849 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/98e27097-4061-4463-b719-bd46a9a030eb) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/14597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/12368 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/2904 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/124184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/167/builds/9768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/157792 "Built successfully") | | 
| | [  ~~🛠 ios-safer-cpp~~](https://ews-build.webkit.org/#/builders/174/builds/937 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/169/builds/11204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/121113 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/19275 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/16194 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/121325 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/31080 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/19284 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/131520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/75104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/16936 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/170/builds/8410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/18890 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/82645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/18620 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/18771 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/18679 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f9212cfe36b8c47b027d8e3f6caa931a8f0d43a8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/42118

smfr committed:
[Async scrolling] Programmatic scrolls can cause missing tiles and offset hit-testing
https://bugs.webkit.org/show_bug.cgi?id=262287
rdar://116205365

Reviewed by Alan Baradlay and Abrar Rahman Protyasha.

This change fixes some long-standing issues with asynchronous scrolling, where
UI and web processes could be left with different scroll positions when both
user and programmatic scrolling was happening at the same time. This could result
in bad tile coverage (missing content), and offset hit-testing.

The fundamental problem was that IPC message for programmatic scrolling, from
web to UI process, could overlap with messages in the other direction for user
scrolling; the messages would pass like ships in the night, with no final
reconciliation step.

307867@main introduced the concept of `ScrollRequestIdentifier`. This change
makes use of those identifiers to ensure that the web process is updated with
the correct scroll position via a message from the UI process back to the web
process. Every programmatic scroll IPC now triggers a response message, which
returns the originating identifier. This allows `RemoteScrollingCoordinator` in
the web process to know when there's programmatic scroll in flight. When there
is, we have to ignore user scrolls, because they will contain a scroll position
that is stale relative to the current web process scroll position. With multiple
programmatic scrolls in a row, we also have to ignore the response for the same
reason. We need to take care to fire `scrollend` when necessary.

ScrollUpdate now has a data member with two variants; one for responses to
programmatic scrolls (`ScrollRequestResponseData`) and one for other messages
from UI to web process.

We actually apply the scroll position in `RemoteScrollingCoordinator::scrollUpdateForNode()`,
to handle the case where the web process sent a "delta" scroll (e.g. for scroll
anchoring or `scrollBy`); the response will contain the most recent UI-side scroll
position with the delta applied.

There was already a `ProgrammaticScrollDidEnd` sent back to the web process; now
that we always send `ScrollRequestResponse` updates back, use that to trigger
`scrollend` via a `ShouldFireScrollEnd` member on ScrollUpdate.
`scrollingTreeNodeDidStopProgrammaticScroll` is no longer needed.

`scrollingTreeNodeRequestsScroll()` has to distinguish between "scrolling tree
handled it for animated scroll", and "scrolling tree delayed handling for
delegated scrolling" so add `RequestsScrollHandling` to indicate that.

The changes in `ScrollingTreeScrollingNode::handleScrollPositionRequest()` and
`RemoteScrollingCoordinatorProxy::adjustMainFrameDelegatedScrollPosition()` are
needed to handle the iOS-specific behavior, seen in `RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction()`,
where the application of the `requestedScroll` is delayed until after layer tree
commit; `didHandleScrollRequestForNode()` needs to be called once the final scroll
position has been computed (this is tested by existing tests).

Tests: fast/scrolling/mac/momentum-then-programmatic-hit-test.html
       fast/scrolling/mac/momentum-then-programmatic-missing-tiles.html

* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-hit-test-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-hit-test.html: Added.
* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-missing-tiles-expected.html: Added.
* LayoutTests/fast/scrolling/mac/momentum-then-programmatic-missing-tiles.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-scroll-anchoring/dirty-contents-reselect-anchor.tentative-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::requestScrollToPosition):
(WebCore::AsyncScrollingCoordinator::synchronizeStateFromScrollingTree):
(WebCore::AsyncScrollingCoordinator::applyScrollPositionUpdate):
(WebCore::AsyncScrollingCoordinator::updateScrollPositionAfterAsyncScroll):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::ScrollUpdate::canMerge const):
(WebCore::ScrollUpdate::merge):
(WebCore::operator<<):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::ScrollUpdate::canMerge const): Deleted.
(WebCore::ScrollUpdate::merge): Deleted.
* Source/WebCore/page/scrolling/ScrollingTree.h:
(WebCore::ScrollingTree::scrollingTreeNodeDidStopWheelEventScroll):
(WebCore::ScrollingTree::scrollingTreeNodeRequestsScroll):
(WebCore::ScrollingTree::didHandleScrollRequestForNode):
(WebCore::ScrollingTree::scrollingTreeNodeDidStopProgrammaticScroll): Deleted.
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.cpp:
(WebCore::ScrollingTreeScrollingNode::handleScrollPositionRequest):
(WebCore::ScrollingTreeScrollingNode::didStopProgrammaticScroll): Deleted.
* Source/WebCore/page/scrolling/ThreadedScrollingTree.cpp:
(WebCore::ThreadedScrollingTree::scrollingTreeNodeRequestsScroll):
(WebCore::ThreadedScrollingTree::scrollingTreeNodeDidScroll):
(WebCore::ThreadedScrollingTree::didHandleScrollRequestForNode):
(WebCore::ThreadedScrollingTree::scrollingTreeNodeScrollUpdated):
* Source/WebCore/page/scrolling/ThreadedScrollingTree.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::adjustMainFrameDelegatedScrollPosition):
(WebKit::RemoteScrollingCoordinatorProxy::sendScrollingTreeNodeUpdate):
(WebKit::RemoteScrollingCoordinatorProxy::scrollingTreeNodeRequestsScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.cpp:
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopAnimatedScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopWheelEventScroll):
(WebKit::RemoteScrollingTree::scrollingTreeNodeRequestsScroll):
(WebKit::RemoteScrollingTree::didHandleScrollRequestForNode):
(WebKit::RemoteScrollingTree::scrollingTreeNodeDidStopProgrammaticScroll): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingTree.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::scrollingTreeNodeRequestsScroll):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingTreeMac.mm:
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopAnimatedScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopWheelEventScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeRequestsScroll):
(WebKit::RemoteScrollingTreeMac::scrollingTreeNodeDidStopProgrammaticScroll): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollingCoordinator.mm:
(WebKit::RemoteScrollingCoordinator::willSendScrollPositionRequest):
(WebKit::RemoteScrollingCoordinator::scrollUpdateForNode):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::updateVisibleContentRects):

Canonical link: https://commits.webkit.org/308215@main
